Land record table

Pi03k opened the pull request:
https://bugs.openmw.org/issues/936
![land_table_ft_936](https://cloud.githubusercontent.com/assets/1756537/14644804/a5ecd952-0653-11e6-9312-773148fdee90.png)

scrawl commented on components/esm/loadland.hpp:
```diff
@@ -110,7 +110,7 @@ struct Land
     };
 
-    void load(ESMReader &esm, bool &isDeleted);
+    virtual void load(ESMReader &esm, bool &isDeleted);
```
Why virtual?

Pi03k replied:
apps/opencs/model/world/land.hpp line 18 has this exact method in derived class. Without virtual this method would be shadowed. Adding virtual will allow for dynamic dispatch / polymorphism / method overriding. Right?

scrawl replied:
Yes, but OpenCS <i>knows</i> that it is dealing with the derived class, so the dynamic dispatch should not be necessary. And as for the game itself, there is no reason for the function to be virtual since it's not being overridden there.

Pi03k replied:
ok. will remove that.

Pi03k replied:
done

zinnschlag commented:
Why the changes to the saving code? Looks random.

Pi03k commented:
Small refactoring. I've removed CSMDoc::WriteLandCollectionStage and replaced in with instantiation of template WriteCollectionStage (one class less - code easier to understand/maintain). WriteLandCollectionStage received reference to document class where it really needs only land collection.

zinnschlag commented:
It seems the perform function for the Land records is functionally identical to the general perform function of the WriteCollectionStage template (except for the scope check which isn't an issue). Not sure how that happened. But if Land saving is refactored we can as well throw out the specialised function entirely.

Pi03k commented:
Should I remove specialised function and leave default template impl?

zinnschlag commented:
Looks like that is the way to go. I suggest to do a quick test run with it afterwards (load, save, test in OpenMW). The save code is crucial for OpenMW-CS and any modification to it (no matter how trivial) should be verified through a test.

Pi03k commented:
It looks that vanilla (master) doesn't save Land structure (with merged initial support by @cc9cii ). With my changes situation is the same. I need to debug and fix that. Save doesn't communicate user about any difficulties. Something weird is going on with saving. I'll describe it later.

zinnschlag commented:
Just tested with master. Loading and saving of land records works fine here.

Pi03k commented:
How did you test that if there is no land table? Could you please describe your testing procedure?

zinnschlag commented:
- Load Morrowind.esm into OpenMW-CS
- Save as Morrowind.omwgame
- Load Morrowind.omwgame into OpenMW-CS -> terrain renders correctly
- Load Morrowind.omwgame into OpenMW -> terrain renders correctly

Pi03k commented:
I've run game with modified omwgame and it didn't crash. Doest it mean that terrain renders correctly?

zinnschlag commented:
> Could you please elaborate more on 'renders correctly'?

Looks like normal terrain. I haven't checked if it matches exactly the original but I didn't notice any differences.

> Is step consisting of conversion from esm to omwgame necessary each time?

OpenMW-CS can not save in esm format.

Pi03k commented:
I've posted my testing of master openmw-cs on forum
https://forum.openmw.org/viewtopic.php?f=7&t=3453

scrawl commented:
For reference, the last comment in said thread:
> 
You added a land record with an invalid name. That needs to be handled in the LAND table creator bar. The only valid IDs are of the format #x y
Instead of entering an ID string it might be better to have two integer fields for input.

Anyway, the LAND code wasn't written for that. I guess you may have trashed the data structure. In fact the esm/omwgame file format can not store LAND records with such an ID. No idea what kind of junk ended up being written to the file.


So the PR is just missing ID validation? Anything else?

akortunov commented:
Feature 936 is implemented. I think we can close this PR now.
